CrazyCanvas/Source/ECS/Systems/Player/WeaponSystem.cpp

```diff
@@ -96,25 +118,66 @@ void WeaponSystem::Tick(LambdaEngine::Timestamp deltaTime)
-		weaponComponent.CurrentCooldown -= onCooldown * dt;
+		const bool hasAmmo		= weaponComponent.CurrentAmmunition > 0;
+		const bool isReloading	= weaponComponent.ReloadClock > 0.0f;
+		if (!hasAmmo && !isReloading)
+		{
+			StartReload(weaponComponent);
+		}
+
+		const bool onCooldown = weaponComponent.CurrentCooldown > 0.0f;
+		if (onCooldown)
+		{
+			weaponComponent.CurrentCooldown -= dt;
+		}
+
+		if (isReloading)
+		{
+			LOG_INFO("Reloading");
+
+			weaponComponent.ReloadClock -= dt;
+			if (weaponComponent.ReloadClock < 0.0f)
+			{
+				weaponComponent.ReloadClock			= 0.0f;
+				weaponComponent.CurrentAmmunition	= 5;
 
-		if (Input::GetMouseState().IsButtonPressed(EMouseButton::MOUSE_BUTTON_FORWARD) && !onCooldown)
+				LOG_INFO("Reload Finish");
+			}
+		}
+
+		// Reload if we are not reloading
+		if (Input::IsKeyDown(EKey::KEY_R) && !isReloading)
+		{
+			StartReload(weaponComponent);
+		}
+		else if (!onCooldown) // If we did not hit the reload try and shoot
 		{
-			const PositionComponent& positionComp = pPositionComponents->GetConstData(playerEntity);
-			const RotationComponent& rotationComp = pRotationComponents->GetConstData(playerEntity);
-			const VelocityComponent& velocityComp = pVelocityComponents->GetConstData(playerEntity);
+			if (Input::GetMouseState().IsButtonPressed(EMouseButton::MOUSE_BUTTON_FORWARD))
+			{
+				const PositionComponent& positionComp = pPositionComponents->GetConstData(playerEntity);
+				const VelocityComponent& velocityComp = pVelocityComponents->GetConstData(playerEntity);
+				const RotationComponent& rotationComp = pRotationComponents->GetConstData(playerEntity);
+
+				TryFire(EAmmoType::AMMO_TYPE_PAINT, weaponComponent, positionComp.Position + glm::vec3(0.0f, 1.0f, 0.0f), rotationComp.Quaternion, velocityComp.Velocity);
+			}
+			else if (Input::GetMouseState().IsButtonPressed(EMouseButton::MOUSE_BUTTON_BACK))
+			{
+				const PositionComponent& positionComp = pPositionComponents->GetConstData(playerEntity);
+				const VelocityComponent& velocityComp = pVelocityComponents->GetConstData(playerEntity);
+				const RotationComponent& rotationComp = pRotationComponents->GetConstData(playerEntity);
 
-			Fire(weaponComponent, positionComp.Position, rotationComp.Quaternion, velocityComp.Velocity);
+				TryFire(EAmmoType::AMMO_TYPE_WATER, weaponComponent, positionComp.Position + glm::vec3(0.0f, 1.0f, 0.0f), rotationComp.Quaternion, velocityComp.Velocity);
+			}
 		}
 	}
 }
 
-void WeaponSystem::Fire(WeaponComponent& weaponComponent, const glm::vec3& playerPos, const glm::quat& direction, const glm::vec3& playerVelocity)
+void WeaponSystem::Fire(EAmmoType ammoType, WeaponComponent& weaponComponent, const glm::vec3& playerPos, const glm::quat& direction, const glm::vec3& playerVelocity)
 {
 	using namespace LambdaEngine;
 
-	weaponComponent.CurrentCooldown = 1.0f / weaponComponent.FireRate;
+	// Tick down ammunition
+	weaponComponent.CurrentAmmunition--;
```

Comment:
What happens if CurrentAmmuition is 0?

Reply:
Well basically infinite ammo since we get 4294967295. Should not be a problem tho since we should not enter this function if the ammunition is less than 1. But your concern is valid and I can put up a quick fix.